### ISSUE TITLE
Convert dump.json into file per sheet [$]

### ISSUE THREAD

**Issue by atifaziz** [audreyt/ethercalc#414]
Wondering what the architectural reason is for creating a single `dump.json`?  Is it simplicity? Why not write each sheet in its own file? Isn't it equally simpler & safer? Would also avoid collateral damage, as in #403.

**Comment by ocdtrekkie** [comment]
dump.json is a fallback method. The preferred solution is to use Redis, but it will use a JSON file if Redis is not present.

**Comment by atifaziz** [comment]
@ocdtrekkie I am aware of that but still wondering if there was another reason for the single dump file besides it being too much trouble to implement sheet-per-file.

**Comment by ocdtrekkie** [comment]
It strikes me as particularly unusual for a web app to store documents in files. o_o Web apps generally store documents as databases. Redis is a database. The JSON file is basically a dump of a database.

Only @audreyt could speak to the reasons why for this being the choice made here, but it seems like sheet-per-file is a particularly unusual setup for this kind of application.

**Comment by audreyt** [comment]
It was just for ease of implementation. I would certainly take a patch that implements sheet-per-file (perhaps using `encodeURIComponent` for the file name).

**Comment by atifaziz** [comment]
@audreyt Thanks. Good to know a PR would be welcome.

@ocdtrekkie A database is a concurrency read/write manager on top of files on disk. Unless EtherCalc is being farmed out across servers for big scale, it can also serve to be that database for a small setup (e.g. departmental) of one server. I'm certainly not advocating this for production setup. I was just thinking about development and sandbox environments.

The benefits of breaking out into files I'm imagining would be:

- Files can be plain text and use the native SocialCalc sheet save format; more readable
- Separation of log, audit and snapshot
- Corruption of one sheet does not affect others
- Shell management of files on disk
- Backup and versioning
- Less resources required to JSON encode and dump the _entire_ database and _all_ sheets irrespective of which ones are actively changing.
- Can scale reasonably for small setups
- Migration to Redis would still be simple

**Comment by atifaziz** [comment]
I just posted a $50 [bounty](https://www.bountysource.com/issues/36409859-convert-dump-json-into-file-per-sheet) :candy: if anyone would like to implement this feature. Hopefully it'll pay for a super expensive :pizza: + :beer: (in some countries you could feed a whole family) as gratitude.

---
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=36409859)](https://www.bountysource.com/issues/36409859-convert-dump-json-into-file-per-sheet)

**Comment by eddyparkinson** [comment]
**Extract a single sheet**

With bash 

```
echo 'Creating sheet:examples'
echo '{' > examples_dump.json

echo '"snapshot-examples":' >> examples_dump.json
json -f ethercalc/dump.json -j snapshot-examples >> examples_dump.json

echo ',' >> examples_dump.json

echo '"snapshot-examples_formdata":' >> examples_dump.json
json -f dump.json -j snapshot-examples_formdata >> examples_dump.json

echo '}' >> examples_dump.json 

```

**Comment by SpaVec** [comment]
I made a pull request, it saves it as a folder containing different files for each sheet and the other database's attributes
https://github.com/audreyt/ethercalc/pull/597

**Comment by atifaziz** [comment]
This was closed via fc031590a147e44eba317d969ce0febcf85c4575, as merge of PR #597. Thanks again, @SpaVec, and don't forget to claim your bounty.
